src/aws_encryption_sdk/keyring/raw_keyring.py

```diff
@@ -0,0 +1,390 @@
+            key_namespace=self.key_namespace, key_name=self.key_name, wrapping_key=self._wrapping_key_structure
+        )
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        """Generate a data key if not present and encrypt it using any available wrapping key
+
+        :param encryption_materials: Encryption materials for the keyring to modify
+        :type encryption_materials: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        :returns: Optionally modified encryption materials
+        :rtype: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        """
+        if not encryption_materials.data_encryption_key:
+            plaintext_data_key = generate_data_key(
+                encryption_materials=encryption_materials, key_provider=self._key_provider
+            )
+        else:
+            plaintext_data_key = encryption_materials.data_encryption_key.data_key
+
+        # Encrypt data key
+        encrypted_wrapped_key = self._wrapping_key_structure.encrypt(
+            plaintext_data_key=plaintext_data_key, encryption_context=encryption_materials.encryption_context
+        )
+
+        # EncryptedData to EncryptedDataKey
+        encrypted_data_key = serialize_wrapped_key(
+            key_provider=self._key_provider,
+            wrapping_algorithm=self._wrapping_algorithm,
+            wrapping_key_id=self.key_name,
+            encrypted_wrapped_key=encrypted_wrapped_key,
+        )
+
+        # Update Keyring Trace
+        if encrypted_data_key:
+            keyring_trace = KeyringTrace(
+                wrapping_key=encrypted_data_key.key_provider, flags={KeyringTraceFlag.WRAPPING_KEY_ENCRYPTED_DATA_KEY}
+            )
+
+            # Add encrypted data key to encryption_materials
+            encryption_materials.add_encrypted_data_key(
+                encrypted_data_key=encrypted_data_key, keyring_trace=keyring_trace
+            )
+        return encryption_materials
+
+    def on_decrypt(self, decryption_materials, encrypted_data_keys):
+        # type: (DecryptionMaterials, Iterable[EncryptedDataKey]) -> DecryptionMaterials
+        """Attempt to decrypt the encrypted data keys.
+
+        :param decryption_materials: Decryption materials for the keyring to modify
+        :type decryption_materials: aws_encryption_sdk.materials_managers.DecryptionMaterials
+        :param encrypted_data_keys: List of encrypted data keys
+        :type: List of `aws_encryption_sdk.structures.EncryptedDataKey`
+        :returns: Optionally modified decryption materials
+        :rtype: aws_encryption_sdk.materials_managers.DecryptionMaterials
+        """
+        if decryption_materials.data_encryption_key:
+            return decryption_materials
+
+        # Decrypt data key
+        expected_key_info_len = len(self._key_info_prefix) + self._wrapping_algorithm.algorithm.iv_len
+        for key in encrypted_data_keys:
+            if (
```

Comment:
To improve the readability of this section, let's invert this logic. Rather than entering a deeper nested logic path if the case is true, `continue` if that case is false.